optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningUtils.java

```diff
@@ -181,4 +187,56 @@ public boolean isFieldADeepCloneProperty(Field field, Class<?> owningClass) {
         }
         return false;
     }
+
+    /**
+     * @return never null
+     */
+    public Set<Class<?>> getDeepClonedTypeArguments(Type genericType) {
+        // Check the generic type arguments of the field.
+        // Yes, it is possible for fields and methods, but not instances!
+        if (genericType instanceof ParameterizedType) {
+            Set<Class<?>> deepClonedTypeArguments = new HashSet<>();
+            ParameterizedType parameterizedType = (ParameterizedType) genericType;
+            for (Type actualTypeArgument : parameterizedType.getActualTypeArguments()) {
+                if (actualTypeArgument instanceof Class
+                        && isClassDeepCloned((Class) actualTypeArgument)) {
+                    deepClonedTypeArguments.add((Class) actualTypeArgument);
+                }
+                deepClonedTypeArguments.addAll(getDeepClonedTypeArguments(actualTypeArgument));
+            }
+            return deepClonedTypeArguments;
+        }
+        return Collections.emptySet();
+    }
+
+    public Set<Class<?>> getDeepClonedClasses(Collection<Class<?>> entitySubclasses) {
+        Set<Class<?>> deepClonedClassSet = new HashSet<>();
+        Set<Class<?>> toSearch = new HashSet<>();
+
+        toSearch.add(solutionDescriptor.getSolutionClass());
+        toSearch.addAll(solutionDescriptor.getEntityClassSet());
+        toSearch.addAll(entitySubclasses);
```

Comment:
This could actually be a `stream.forEach()` - creating and resizing this collection (twice) is pure waste.